### PR TITLE
Introduce twitter_timeline liquid tag (#7605)

### LINKED ISSUE #7605: Twitter timeline support
**Describe the bug**

Twitter timelines, such as [this one](https://twitter.com/FreyaHolmer/timelines/1215413954505297922), are not supported by the liquid tags.
Instead, a "Liquid error: internal" message is shown.

**To Reproduce**

{% twitter 1215413954505297922 %}

**Expected behavior**

Either a message such as "Timelines are not supported", or a proper "embed" for these.

I'm not sure how the API handles these results or if we can actually embed them, but the website should beautifully handle this case.

### PR DESCRIPTION
**Pull request by rafaltrojanowski**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [x] Documentation Update

## Description

This PR introduces a new twitter_timeline liquid tag. See #7605

## Related Tickets & Documents

Closes #7605

## QA Instructions, Screenshots, Recordings

1. Go to /new
2. Create article with following content:
```{% twitter_timeline https://twitter.com/FreyaHolmer/timelines/1215413954505297922 %}```
3. Click 'Preview'
3. Click 'Save'.

Screen Shot 2563-10-15 at 11 45 39

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [x] no documentation needed

alt text

### CODE REVIEW

**Review by rhymes — COMMENTED** [review]
Hi @rafaltrojanowski thanks for kickstarting this. 

A few notes:

1. I can't see it in Firefox 82, mostly because the browser's builtin content blocking blocks it:

> The resource at “https://platform.twitter.com/widgets.js” was blocked because content blocking is enabled

https://developer.mozilla.org/en-US/docs/Mozilla/Firefox/Privacy/Tracking_Protection

I tried to disable the "Enhance tracking protection" feature for the website, and it worked, but we have to be aware that Firefox users may never see the timeline and not know they aren't seeing it. Is there an option B to render the timeline?

1. The timeline widget is too long, this is as long as the timeline on Twitter, we might need to add scrollbars otherwise the vertical size of the article can go on forever. An example:

Screenshot_2020-10-13 test twitter timeline - DEV(local) Community 👩‍💻👨‍💻

cc @ludwiczakpawel

File: app/liquid_tags/twitter_timeline_tag.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+class TwitterTimelineTag < LiquidTagBase
+  include ActionView::Helpers::AssetTagHelper
+  PARTIAL = "liquids/twitter_timeline".freeze
+
+  URL_REGEXP = %r{\Ahttps://twitter\.com/[a-zA-Z0-9]+/timelines/\d+\Z}.freeze
+
+  SCRIPT = <<~JAVASCRIPT.freeze
+    <script src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>
```
[the remark this answers is not in the document]
Comment (benhalpern) [reply app/liquid_tags/twitter_timeline_tag.rb:8]: Yeah. We wouldn't want this to be a _blocking_ script.

File: app/liquid_tags/twitter_timeline_tag.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    raise_error unless valid_link?(href)
+    href
+  end
+
+  def valid_link?(link)
+    (link =~ URL_REGEXP)&.zero?
+  end
+
+  def raise_error
+    raise StandardError, "Invalid Twitter timeline URL"
```
[the remark this answers is not in the document]
Comment (rafaltrojanowski) [reply app/liquid_tags/twitter_timeline_tag.rb:42]: Good point 👍

File: app/liquid_tags/twitter_timeline_tag.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (rafaltrojanowski) [reply app/liquid_tags/twitter_timeline_tag.rb:8]: Should we use `async` or `defer` for it?

Also, I probably have forgotten to place it in 'forem/app/views/comments/index.html.erb' (https://github.com/forem/forem/issues/7605#issuecomment-707097018)

File: app/javascript/article-form/components/Preview.jsx [hunk 3]
```diff
@@ -68,6 +69,16 @@ const previewResponsePropTypes = PropTypes.shape({
...
 export const Preview = ({ previewResponse, articleState, errors }) => {
+  useEffect(() => {
+    const script = document.createElement('script');
+    script.src = 'https://platform.twitter.com/widgets.js';
+    script.async = true;
+    document.body.appendChild(script);
+    return () => {
+      document.body.removeChild(script);
+    };
+  }, []);
```
[the remark this answers is not in the document]
Comment (rafaltrojanowski) [reply app/javascript/article-form/components/Preview.jsx:80]: Right, I'm trying to figure out how to do it :)

As far as I know, there is no info in the `previewResponse` object about used liquid tags.

**Comment by ludwiczakpawel** [comment]
Hey @rafaltrojanowski (cześć! 🇵🇱)

It does seem like we need to limit the height of this liquid tag to avoid situations thatx @rhymes mentioned. I think the most sensible way to do it is to go with either `overflow-y: auto` OR some fadeout effect similar to what we have in github liquid tag or long comments in feed:

### Github liquid tag fadeout ([source](https://dev.to/devteam/spam-sucks-dm8)):

image

### Comments fadeout:

image

It would be also good to wrap this entire block with `.crayons-card` container so it looks more... umm.. wrapped..? :)

**Comment by rafaltrojanowski** [comment]
Cześć @ludwiczakpawel 

I've added styling updates based on Github liquid tag, please have a look:

Screen Shot 2563-10-15 at 11 45 39

**Review by ludwiczakpawel — APPROVED** [review]
Cheers, looking good from my perspective!

**Review by nickytonline — APPROVED** [review]
Tested this out locally and it works great! Thanks so much for the pull request @rafaltrojanowski! 🚀

**Comment by rafaltrojanowski** [comment]
Hey @rhymes 

Unfortunately, I didn't have a chance to look into the problem on Firefox yet, maybe it could be tacked in a separate PR?
Currently, I don't have an alternative solution in mind.

File: app/liquid_tags/twitter_timeline_tag.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  private
+
+  def parse_link(link)
+    href = ActionController::Base.helpers.strip_tags(link).strip
+    raise_error unless valid_link?(href)
+    href
+  end
+
+  def valid_link?(link)
+    (link =~ URL_REGEXP)&.zero?
```
Comment (citizen428) [inline app/liquid_tags/twitter_timeline_tag.rb:38]: **suggestion (non-blocking):** We can use the `match?` predicate method here.

```suggestion
    link.match?(URL_REGEXP)
```

**Review by citizen428 — APPROVED** [review]

**Pull request merged**